pkg/dependencies/factory.go

```diff
@@ -580,6 +582,42 @@ func (f *Factory) WithProvider(clusterConfigFile string, clusterConfig *v1alpha1
 	return f
 }
 
+// WithKubeconfigWriter adds the KubeconfigReader dependency depending on the provider.
+func (f *Factory) WithKubeconfigWriter(clusterConfig *v1alpha1.Cluster) *Factory {
+	f.WithUnAuthKubeClient()
+	f.buildSteps = append(f.buildSteps, func(ctx context.Context) error {
+		if f.dependencies.KubeconfigWriter != nil {
+			return nil
+		}
+		writer := kubeconfig.NewClusterAPIKubeconfigSecretWriter(f.dependencies.UnAuthKubeClient)
+		switch clusterConfig.Spec.DatacenterRef.Kind {
+		case v1alpha1.DockerDatacenterKind:
+			f.WithDocker()
```

Comment:
you need to call `WithDocker` from outside of the build step

the factory more or less works like this

1. Define deps
1.1 Your Deps define their own deps
2. Traverse the tree that has been declared by the dependencies, building them one by one. This uses DFS, which guarantees an object dependencies are built before itself

If you define the deps during the build process is like not declaring them at all

Let me know if this requires a lengthier explanation and we can chat offline. This can be tricky and it's important to get it right.